rskj-core/src/main/java/co/rsk/core/bc/MiningMainchainViewImpl.java

```diff
@@ -73,24 +79,84 @@ public void addBest(BlockHeader bestHeader) {
         }
     }
 
-    private void fillInternalStore(BlockHeader bestHeader) {
-        List<BlockHeader> newHeaderChain = new ArrayList<>(height);
-        BlockHeader currentHeader = bestHeader;
-        for(int i = 0; i < height; i++) {
-            newHeaderChain.add(currentHeader);
+    /**
+     * Given a new best header and the index of its parent rebuild the mainchain using the index as the pivot point
+     * by discarding all headers subsequent to the it, setting the new header as the tip and refilling with as many
+     * are needed to complete the desired height
+     *
+     * @param bestHeader The best header to be on top of the chain
+     * @param parentIndex List index of the best header's parent
+     */
+    private void addBestAndRebuildFromParent(BlockHeader bestHeader, int parentIndex) {
+        List<BlockHeader> staleHeaders = mainchain.stream()
+                .limit(parentIndex)
+                .collect(Collectors.toList());
+
+        List<BlockHeader> commonAncestorChain = mainchain.stream()
+                .skip(staleHeaders.size())
+                .collect(Collectors.toList());
+
+        List<BlockHeader> newMainchain = Stream.concat(
+                Arrays.asList(bestHeader).stream(),
+                commonAncestorChain.stream())
+                .collect(Collectors.toList());
+
+        mainchain = newMainchain;
+
+        fillMissingHeaders();
+        deleteHeaders(staleHeaders);
```

Comment:
Why do you use `deleteHeaders` here if after `deleteEntriesOutOfBoundaries` is called ? 
I understand these methods do the same.

Reply:
Indeed, the methods do more or less the same, but the former is used to delete the stale headers after the chain is rebuilt based on a specific parent from the chain. However, these same stale headers would be eventually removed by `deleteEntriesOutOfBoundaries`.

I'll change it to leave the headers deletion to the later method since I doubt there'd be memory issues.